bf: ZENKO-1024 fail retries metrics

philipyoo opened the pull request:
Quick fix for backbeat failure/pending metrics and failure object metrics:
- Failures (bucket-level) metrics should represent list of failures object keys
- Retry Fails GET listings and POST requests should filter out duplicate version ids 
   as they should be unique.
- Any duplicates found should try to normalize pending metrics by first filtering from
   the list, attempt to remove the duplicate keys from redis, and then increment
   pending metrics

To be merged AFTER https://github.com/scality/backbeat/pull/415

To be merged with https://github.com/scality/Arsenal/pull/557

philipyoo committed:
bf: ZENKO-1024 get unique version ids retry list

For fail retry requests:
GET requests, after getting redis keys, check and filter any
duplicate version id's in list. Also add debug log.
POST requests, after parsing user request body, check and
filter any duplicate version id's the user may have made
in the request.

philipyoo committed:
bf: ZENKO-1024 replace failures metrics

Failures metrics will use list of object failure keys
instead of the reported failures from the replication
status processor

philipyoo committed:
bf: ZENKO-1024 try to normalize pending counter

When failure keys with identical version ids are found,
we first filter from the original request, but then we
also attempt to flush this bad data.

We want to do this in an attempt to reduce duplicate keys,
and to try and fix pending global metrics.

bert-e commented:
# Hello philipyoo,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @vrancurel

* @rahulreddy

* @bennettbuchanan

* @jonathan-gramain

philipyoo commented:
@bert-e wait

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+            if (err) {
+                this._logger.error('error getting metric: failures', {
+                    error: err,
+                    method: 'BackbeatAPI.getFailMetrics',
```
~getFailMetrics~ getFailedMetrics

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+            //           <versionID>, <location>, <timestamp>]
+            const parts = key.split(':');
+            const versionID = process.env.CI === 'true' ? parts[6] : parts[5];
+            if (versionID === '') {
```
I guess this is the case for non-versioned objects?

philipyoo replied:
Ah, I did this to fix a test case. Looking at the redis constants, maybe I can match part size instead

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+            }
+            // if the versionID already exists, debug log because an error
+            // occurred somewhere
+            if (uniqVersionIDs[versionID]) {
```
I'm wondering if there is an off-chance that there could be multiple version IDs across buckets and objects. Perhaps to be safe, we should check that the bucket, object, and versionId is unique. In that case, we could just update the hash key to include all of those values.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+                this._logger.debug('duplicate version id was found in failed ' +
+                'object redis keys', {
+                    method: 'BackbeatAPI._filterFailObjectKeys',
+                    message: 'Known bug with unknown origin. Debug and ' +
```
I'm thinking the debug log message is enough here and instead add the sentiment of this message as a comment. For example, I can imagine a user seeing this and wondering "who, exactly, should I report the behavior to?".

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+                uniqVersionIDs[versionID] = key;
+            }
+        });
+        filteredKeys.push(...Object.keys(uniqVersionIDs)
```
I think we can avoid this loop by just pushing the key to `filteredKeys` in the `else` block above.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+    _filterFailObjectKeys(keys) {
+        const filteredKeys = [];
+        const uniqVersionIDs = {};
+        const removeKeys = [];
```
I think `duplicateKeys` might be more expressive.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+                    return done(null, entry);
+                }),
+                (entry, done) => {
+                    if (!entry) {
```
I'm not sure I understand when this would this occur.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,156 @@ class BackbeatAPI {
+                        method: 'BackbeatAPI._normalizeFailMetrics',
+                        key,
+                    });
+                    return next();
```
I think we want to return the error here if there is one.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -636,6 +793,19 @@ class BackbeatAPI {
                 }
+                // NOTE: if multiple keys, an error occurred
+                // Filter/Remove duplicate object keys with same version id
+                if (keys.length > 1) {
```
👍

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -705,6 +875,22 @@ class BackbeatAPI {
+        reqBody = reqBody.filter(entry => {
+            if (!entry.VersionId) {
+                return true;
+            } else if (!trackVersionIDs[entry.VersionId]) {
```
As mentioned in another comment, I think it could be worthwhile to also check the bucket and object names here.

Minor: Let's make this another `if` block (rather than `else..if`) because we return in the first one.

bennettbuchanan reviewed (commented):
Thanks for getting this! It looks like a good way to address the issue we were seeing where there were duplicate failure entries of the same object-key-versionId. I think this will help avoid that case in the future.

JianqinWang commented on tests/functional/api/BackbeatServer.js:
```diff
@@ -263,7 +263,20 @@ describe('Backbeat Server', () => {
             statsClient.reportNewRequest(`${site2}:${BYTES_DONE}`, 1874);
             statsClient.reportNewRequest(`${site2}:${BYTES_FAIL}`, 575);
+
+            // TODO: quick fix made for failures/pending
```
? is a quick fix needed or is this an old TODO that has been completed in this commit

philipyoo replied:
Yup, I want to keep this todo. I want to improve these tests, but currently not sure how atm

JianqinWang replied:
Can you update the TODO to be more descriptive of future work? maybe something like 'refactor/improve tests for failures/pending, current tests are modified as a quick fix'. Sorry for being so picky, the current `TODO` sounds more like 'here's what happened' than a 'this is what should happen'

JianqinWang commented on tests/functional/api/BackbeatServer.js:
```diff
@@ -543,9 +556,9 @@ describe('Backbeat Server', () => {
                 assert(res.failures.description);
                 // Failures count = OPS_FAIL
-                assert.equal(res.failures.results.count, 150);
+                assert.equal(res.failures.results.count, 1);
```
the values for these comparison have been changed (no longer matching the values set OPS_FAIL and BYTES_FAIL in `before`), perhaps comment needs updating?

jonathan-gramain commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,150 @@ class BackbeatAPI {
+        }
+        async.eachLimit(keys, 10, (key, next) => {
+            async.waterfall([
+                done => this._redisClient.batch([['get', key]], (err, res) => {
```
Since we are not targeting efficiency I think it's fine as it is, but we could also have used a single batch to get all keys.

jonathan-gramain commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,150 @@ class BackbeatAPI {
+                        // error occurred, skip for next time
+                        return done(cmdErr);
+                    }
+                    if (!sourceRole) {
```
I don't get the link between `sourceRole` missing and the key being deleted. If this is correct, can you comment to explain why?

philipyoo replied:
When making the `get` to redis, if no key was found in redis, `sourceRole` will be null

jonathan-gramain replied:
Yes, but why `sourceRole`? Is it the value stored in the key?

jonathan-gramain replied:
Or could be the argument needed to create the `ObjectFailureEntry`

philipyoo replied:
yes, `sourceRole` is the arn saved in the key

jonathan-gramain commented on lib/api/BackbeatAPI.js:
```diff
@@ -615,6 +622,150 @@ class BackbeatAPI {
+        keys.forEach(key => {
+            // 8 parts: [bb, crr, failed, <bucket>, <key>,
+            //           <versionID>, <location>, <timestamp>]
+            const parts = process.env.CI === 'true' ?
```
I'm not fond of having to do this specific case for the CI as it brings discrepancy between testing and production, is there an easy alternative to avoid it? E.g. are tests using an extra field that cannot be put into the production keys?

philipyoo replied:
there is an additional section for the failed test keys. I tried to match the length in the key and it broke many of the tests. Given lack of time, I decided to keep things as is. But definitely, I'd like to avoid doing this

jonathan-gramain commented on lib/api/BackbeatAPI.js:
```diff
@@ -705,6 +867,23 @@ class BackbeatAPI {
+        // filter duplicate version ids if any
+        const uniqueObjs = {};
+        if (process.env.CI === 'true') {
+            // ignore filtering for tests
```
Why do tests need to ignore filtering?

philipyoo replied:
failure keys rely on valid bucket, key, version ids. I wasn't fully sure how the test mock metadata worked, but the version id's in the failure keys for tests use the same mock metadata. I'd like to eventually adjust the tests, and left comment here https://github.com/scality/backbeat/pull/416/files/df1c6700ef3c549deebdfe3a6b0b6a7a2dd8cbdf#diff-8c901706f0bfac9a0fd4f1d1205a4924R267

I've mainly relied on manual testing for these fixes since the bug reported mainly has to do with user usage cases or timing

jonathan-gramain replied:
OK, maybe worth commenting with a TODO in the code as it does not come obvious.

jonathan-gramain reviewed (commented):
The method looks good to me, I have a few random comments though

philipyoo committed:
bf: ZENKO-1024 add lock for multiple requests

If multiple retry requests are made, to avoid sending
multiple kafka entries for the same keys, we need to
somehow lock the entry. For now, add an additional
redis key to indicate a given failed key is already
being processed by another request.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.0`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of Release Engineering will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of Release Engineering now.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -662,6 +848,37 @@ class BackbeatAPI {
+                return this._redisClient.batch([['set', key, 'true']], err => {
+                    if (err) {
+                        return cb(err);
+                    }
```
Looks like we also need to check for the command error here.

bennettbuchanan commented on lib/api/BackbeatAPI.js:
```diff
@@ -662,6 +848,37 @@ class BackbeatAPI {
+     * @return {undefined}
+     */
+    _applyRetryLockKey(key, cb) {
+        this._redisClient.batch([['get', key]], (err, res) => {
```
Minor: using the redis command 'exists' might be a bit cleaner.

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.0`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue ZENKO-1024.

Goodbye philipyoo.
